master/pkg/stream/stream.go

```diff
@@ -1,41 +1,81 @@
 package stream
 
 import (
+	"encoding/json"
 	"sync"
-
-	"github.com/gorilla/websocket"
 )
 
+// UpsertFunc is a function that overrides the default upsert
+type UpsertFunc func(interface{}) interface{}
+
+// DeleteFunc is a function that overrides the default deletion
+type DeleteFunc func(string, string) interface{}
+
 // Msg is an object with a message and a sequence number and json marshal caching.
 type Msg interface {
 	SeqNum() int64
-	UpsertMsg() *websocket.PreparedMessage
-	DeleteMsg() *websocket.PreparedMessage
+	UpsertMsg() UpsertMsg
+	DeleteMsg() DeleteMsg
 }
 
 // Event contains the old and new version a Msg.  Inserts will have Before==nil, deletions will
 // have After==nil.
 type Event[T Msg] struct {
 	Before *T `json:"before"`
 	After  *T `json:"after"`
+
+	upsertCache interface{}
+	deleteCache interface{}
+}
+
+type PreparableMessage interface{}
+
+type UpsertMsg struct {
+	PreparableMessage
+	JsonKey string
+	Msg     Msg
+}
+
+func (u *UpsertMsg) MarshalJSON() ([]byte, error) {
+	data := map[string]Msg{
+		u.JsonKey: u.Msg,
+	}
+	return json.Marshal(data)
+}
+
+type DeleteMsg struct {
+	PreparableMessage
+	Key     string
+	Deleted string
+}
+
+func (d *DeleteMsg) MarshalJSON() ([]byte, error) {
+	data := map[string]string{
+		d.Key: d.Deleted,
+	}
+	return json.Marshal(data)
 }
 
 // Streamer aggregates many events and wakeups into a single slice of pre-marshaled messages.
 // One streamer may be associated with many Subscription[T]'s, but it should only have at most one
 // Subscription per type T.  One Streamer is intended to belong to one websocket connection.
 type Streamer struct {
 	Cond *sync.Cond
-	// Msgs are pre-marshaled messages to send to the streaming client.
-	Msgs []*websocket.PreparedMessage
+	// Msgs are pre-marshalled messages to send to the streaming client.
```

Comment:
hey now, I might be a terrible speller but `marshaled` is definitely the correct American English spelling

Reply:
(I have to cherish this because people are always fixing my spelling errors)

Reply:
lol ok I'll revert this asap